src/Chapter3.hs

```diff
@@ -1007,6 +1190,162 @@ contestants, and write a function that decides the outcome of a fight!
 -}
 
 
+--             --
+-- Description --
+--             --
+
+newtype Power = Power Int deriving (Eq, Show)
+newtype Def   = Def   Int deriving (Eq, Show)  -- Avoid conflict with Defense
+newtype HP    = HP    Int deriving (Eq, Show)
+
+data Status = Alive | Dead deriving (Show, Eq)
+
+class Fighter a where
+    -- | Get all actions
+    getActions :: a -> [Action]
+    -- | Get current life
+    getHP      :: a -> HP
+    -- | Get current protection
+    getDef     :: a -> Def
+    -- | Get current power when hitting
+    getPower   :: a -> Power
+    -- | Handle impact of damages from another fighter
+    takeHit :: a -> Power -> a
+    -- | Handle life changes
+    heal :: a -> HP -> a
+    -- | Get current status
+    getStatus :: a -> Status
+
+
+-- Avoid collision by appending `'` to previous ones
+-- | A creature that can attack and well ... attack
+data Monster = Monster
+    { mHP :: HP
+    , mPower :: Power
+    , mActions :: [Action]
+    } deriving Show
+
+instance Fighter Monster where
+    getActions m = cycle (mActions m)
+    getHP        = mHP
+    getDef     _ = Def 0
+    getPower     = mPower
+    heal m (HP hp) = m {mHP = HP (currentHp + hp) }
+        where
+            (HP currentHp) = getHP m
+
+    takeHit  m (Power p) = m {mHP = HP (hp - max 0 (p - def))}
+        where
+            -- ?? Better way to write it ??
+            (Def def) = getDef m
+            (HP hp)   = getHP m
+
+    getStatus m = if hp < 1 then Dead else Alive
+        where
+            (HP hp) = getHP m
+
+
+-- | A guy that can attack and defend himself
+data Knight = Knight
+    { kHP :: HP
+    , kDef :: Def
+    , kPower :: Power
+    , kActions :: [Action]
+    } deriving Show
+
+instance Fighter Knight where
+    getActions k = cycle (kActions k)
+    getHP        = kHP
+    getDef       = kDef
+    getPower     = kPower
+    heal m (HP hp) = m {kHP = HP (currentHp + hp) }
+        where
+            (HP currentHp) = getHP m
+
+    takeHit  k (Power p) = k {kHP = HP (hp - max 0 (p - def))}
+        where
+            -- ?? Better way to write it ??
+            (Def def) = getDef k
+            (HP hp)   = getHP k
+
+    -- ?? Possible to avoid same definition for both Knight and Monster ??
+    getStatus m = if hp < 1 then Dead else Alive
+        where
+            (HP hp) = getHP m
+
+
+-- | What possible action a Fighter can do !
+data Action = Hit | Run | Drink Beverage | Dance | Cry deriving (Show, Eq)
```

Comment:
Ideally, it would be nice to have two different types of actions: `MonsterAction` and `KnightAction`, so you can't mix actions of different types for them who can't perform such actions

Reply:
Maybe, but how could I have common actions for both in this case ?

Reply:
@JeremyBois I don't think you need a common action. I suggested having two separate data types exactly because it won't be possible to mix them up. I can see how it makes the current design challenging to change. But this is an open-ended task. and it has multiple ways to solve 🙂